File: android/app/src/main/java/tw/speedalert/speech/VoicePack.kt
package tw.speedalert.speech

import android.content.Context
import android.media.MediaPlayer
import java.io.File

/**
 * Modular prerecorded voice pack — mirrors iOS `VoicePack`.
 * Clips live in `assets/VoicePack/<token>.<ext>`. Sentences are assembled from
 * tokens; if any clip is missing, [play] returns false (caller falls back to TTS).
 */
class VoicePack(private val context: Context) {

    private val exts = listOf("m4a", "caf", "mp3", "wav", "aiff", "aif")
    private val players = ArrayList<MediaPlayer>()

    private fun assetPath(token: String): String? {
        for (e in exts) {
            val p = "VoicePack/$token.$e"
            try {
                context.assets.open(p).close()
                return p
            } catch (_: Exception) { /* keep looking */ }
        }
        return null
    }

    val isReady: Boolean get() = assetPath("lead_front") != null

    /** Copy asset to cache (works even if assets are compressed). */
    private fun fileFor(assetPath: String): File {
        val out = File(context.cacheDir, "VoicePack/" + assetPath.substringAfterLast('/'))
        if (!out.exists()) {
            out.parentFile?.mkdirs()
            context.assets.open(assetPath).use { input ->
                out.outputStream().use { input.copyTo(it) }
            }
        }
        return out
    }

    /** Play [tokens] in order. Returns false if any clip is missing. */
    fun play(tokens: List<String>): Boolean {
        if (tokens.isEmpty()) return false
        val paths = ArrayList<String>(tokens.size)
        for (t in tokens) paths.add(assetPath(t) ?: return false)

        stop()
        playFrom(paths, 0)
        return true
    }

    private fun playFrom(paths: List<String>, index: Int) {
        if (index >= paths.size) return
        val mp = MediaPlayer()
        try {
            mp.setDataSource(fileFor(paths[index]).absolutePath)
            mp.setOnCompletionListener {
                it.release()
                players.remove(it)
                playFrom(paths, index + 1)
            }
            mp.prepare()
            mp.start()
            players.add(mp)
        } catch (_: Exception) {
            runCatching { mp.release() }
        }
    }

    fun stop() {
        for (p in players) {
            runCatching { p.stop() }
            runCatching { p.release() }
        }
        players.clear()
    }
}

File: ios/SpeedAlert/Audio/VoicePack.swift
import Foundation
import AVFoundation

/// Modular **prerecorded voice pack**.
///
/// Sentences are assembled from short clips so a voice actor only records a
/// small set of tokens (see VOICEPACK.md). Clips are looked up in
/// `Documents/VoicePack/` first (drop-in, no rebuild), then the bundled
/// `VoicePack/` resource dir. If any clip is missing, `play(_:)` returns false
/// and the caller falls back to system TTS.
final class VoicePack {
    static let shared = VoicePack()

    private var searchDirs: [URL] = []
    private var players: [AVAudioPlayer] = []
    private let exts = ["m4a", "caf", "mp3", "wav", "aiff", "aif"]

    private init() {
        if let docs = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first {
            searchDirs.append(docs.appendingPathComponent("VoicePack", isDirectory: true))
        }
        if let res = Bundle.main.resourceURL?.appendingPathComponent("VoicePack", isDirectory: true) {
            searchDirs.append(res)
        }
    }

    /// Directory currently providing clips (Documents first, else bundle).
    var activeDir: URL? {
        searchDirs.first { FileManager.default.fileExists(atPath: $0.path) }
    }

    func url(for token: String) -> URL? {
        for dir in searchDirs {
            for e in exts {
                let u = dir.appendingPathComponent("\(token).\(e)")
                if FileManager.default.fileExists(atPath: u.path) { return u }
            }
        }
        return nil
    }

    /// A pack is usable once the anchor clip exists.
    var isReady: Bool { url(for: "lead_front") != nil }

    /// Play `tokens` in order. Returns false if any clip is missing.
    @discardableResult
    func play(_ tokens: [String]) -> Bool {
        guard !tokens.isEmpty else { return false }
        var made: [AVAudioPlayer] = []
        for t in tokens {
            guard let u = url(for: t), let p = try? AVAudioPlayer(contentsOf: u) else { return false }
            p.prepareToPlay()
            made.append(p)
        }
        players = made                     // retain so playback survives
        var offset: TimeInterval = 0
        for p in players {
            p.play(atTime: p.deviceCurrentTime + offset)
            offset += p.duration + 0.03
        }
        return true
    }

    func stop() {
        players.forEach { $0.stop() }
        players.removeAll()
    }
}
